File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

def DFS(x, y):
    global cnt

    cnt += 1
    board[x][y] = 0

    for k in range(4):
        xx = x + dx[k]
        yy = y + dy[k]
        if 0 <= xx < n and 0 <= yy < n and board[xx][yy] == 1:
            DFS(xx, yy)


if __name__ == "__main__":
    n = int(input())
    board = [list(map(int, sys.stdin.readline().rstrip())) for _ in range(n)]
    res = []

    for i in range(n):
        for j in range(n):
            if board[i][j] == 1:
                cnt = 0
                DFS(i, j)
                res.append(cnt)
    res.sort()
    print(len(res))
    for x in res:
        print(x)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.*;
import java.util.ArrayList;
import java.util.Collections;

class Main {
    int[] dx = {-1, 0, 1, 0};
    int[] dy = {0, 1, 0, -1};

    static int n;
    static int[][] board;
    static ArrayList<Integer> list = new ArrayList<>();
    static int cnt;

    public void DFS(int x, int y) {
        cnt++;

        board[x][y] = 0;

        for (int i = 0; i < 4; i++) {
            int xx = x + dx[i];
            int yy = y + dy[i];
            if (0 <= xx && xx < n && 0 <= yy && yy < n && board[xx][yy] == 1) {
                DFS(xx, yy);
            }
        }

    }

    public static void main(String[] args) throws IOException {
        Main T = new Main();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        n = Integer.parseInt(br.readLine());

        board = new int[n][n];

        for (int i = 0; i < n; i++) {
            String[] tmp = br.readLine().split("");
            for (int j = 0; j < n; j++) {
                board[i][j] = Integer.parseInt(tmp[j]);
            }
        }

        int res = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 1) {
                    cnt = 0;
                    res++;
                    T.DFS(i, j);
                    list.add(cnt);
                }
            }
        }

        Collections.sort(list);
        bw.write(res + "\n");
        for (int x : list) {
            bw.write(x + "\n");
        }

        bw.flush();
        bw.close();
    }
}